### DISCUSSION TITLE
20250703

### DISCUSSION THREAD

**Discussion by mipypf** [sinchir0/make_llm_book#12] [General]
hogehoge

**Comment by mipypf** [comment]
4.5 TransformerブロックでAttention層と線形層を接続する
ショートカット接続で勾配消失問題が良化する現象をlayerごとに可視化できて分かりやすかった

4.6 GPTモデルを構築する
https://arxiv.org/pdf/2002.04745
書籍に記載のある順番が良い理由なども言及されている

次回は4.7から

**Comment by sinchir0** [comment]
- arxivではなくar5ivを使うと日本語翻訳がしやすい。
- ショートカット接続を使うと、勾配が小さくなる勾配消失問題を解決できるようになる。
- Transformerブロックはこれまでの集大成

**Comment by eikichi838** [comment]
ショートカット接続は勾配消失問題を解決する。
TransformerBlock内でPre-LayerNormがスタンダード。昔はPost-LayerNormだったが学習のダイナミクスに悪影響を及ぼすらしい。
https://arxiv.org/pdf/2002.04745

**Comment by DM99999** [comment]
P118

スキップ接続すると勾配が消失してないすごい！！

入力に近いほど勾配消失がひどくなるらしい

```python
layer_sizes = [3, 3, 3, 3, 3, 1]  

sample_input = torch.tensor([[1., 0., -1.]])

torch.manual_seed(123)
model_without_shortcut = ExampleDeepNeuralNetwork(
    layer_sizes, use_shortcut=False
)
print_gradients(model_without_shortcut, sample_input)
```

> `layers.0.0.weight has gradient mean of 0.00020173584925942123
layers.1.0.weight has gradient mean of 0.00012011159560643137
layers.2.0.weight has gradient mean of 0.0007152040489017963
layers.3.0.weight has gradient mean of 0.0013988736318424344
layers.4.0.weight has gradient mean of 0.005049645435065031`
> 

```python
torch.manual_seed(123)
model_with_shortcut = ExampleDeepNeuralNetwork(
    layer_sizes, use_shortcut=True
)
print_gradients(model_with_shortcut, sample_input)
```

> `layers.0.0.weight has gradient mean of 0.22169791162014008
layers.1.0.weight has gradient mean of 0.20694105327129364
layers.2.0.weight has gradient mean of 0.32896995544433594
layers.3.0.weight has gradient mean of 0.2665732204914093
layers.4.0.weight has gradient mean of 1.3258540630340576`
> 


P120

Pre-LayerNormを使う

Post-LayerNormはしばしば学習の「ダイナミクス」に悪影響を及ぼす？

> o4-miniより
> 
> - **「学習のダイナミクス」** とは、**勾配の伝播**や**損失の減少速度**、**収束の安定性**など、モデルがどのようにパラメータを更新しながら学習していくかという挙動全般を指します。
> - Post-LayerNorm は、その設計ゆえに勾配が層を通過するたびに乱れやすく、結果として安定した・高速な学習を妨げるため、「ダイナミクスに悪影響を及ぼす」と言われます。
> - 実際の大規模モデルでは、より安定的に学習できる Pre-LayerNorm（あるいは最近注目されている Peri-LayerNorm）への置き換えが増えています。

P121

Transformerブロックの出力では入力の次元が維持されている

この設計により、Transformerブロックを幅広いSequence-to-Sequenceタスクにうまく適用できる
